packages/runner-ct/src/app/SpecList/useFuzzySort.ts

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react'
+import fuzzysort from 'fuzzysort'
+
+interface UseFuzzySort<Input, Result> {
+  search: string
+  items: Input[]
+  options: Fuzzysort.KeyOptions
+  transformResult: (spec: Input, indexes: number[]) => Result
+}
+
+export const useFuzzySort = <T, R>({ search, items, transformResult, options }: UseFuzzySort<T, R>) => {
+  const [lastSearchInput, setLastSearchInput] = React.useState<string>(search)
+  const [cachedFuzzySort, setCachedFuzzySort] = React.useState<R[]>([])
+  const [searchPromise, setSearchPromise] = React.useState<Fuzzysort.CancelablePromise<Fuzzysort.KeyResults<T>>>()
+
+  React.useLayoutEffect(() => {
+    // If no search was entered we just show all the items.
+    if (!lastSearchInput) {
+      return setCachedFuzzySort(items.map((spec) => transformResult(spec, [])))
+    }
+
+    const promise = fuzzysort.goAsync(lastSearchInput, items, options)
+
+    setSearchPromise(promise)
+    promise.then((result) => {
+      return setCachedFuzzySort(result.map((res) => {
+        return transformResult(res.obj, res.indexes)
+      }))
```

Comment:
I would usually drop a nit here about `async/await`, but in this scenario (a React hook), using `async/await` is a little more messy, so this is fine

Reply:
Yeah.... tried async/await but it was a little messier. I think this is more readable.